include/ofi_signal.h

```diff
@@ -87,19 +104,73 @@ static inline void fd_signal_free(struct fd_signal *signal)
 static inline void fd_signal_set(struct fd_signal *signal)
 {
 	char c = 0;
-	if (signal->wcnt == signal->rcnt) {
-		if (ofi_write_socket(signal->fd[FI_WRITE_FD], &c, sizeof c) == sizeof c)
-			signal->wcnt++;
+	bool cas; /* cas result */
+	int write_rc;
+
+	cas = ofi_atomic_cas_bool_strong32(&signal->state,
+					   OFI_SIGNAL_UNSET,
+					   OFI_SIGNAL_WRITE_PREPARE);
```

Comment:
We don't need to change this patch, but fd_signal_set may frequently be made with the signal state == SET.  If atomic_get is significantly faster than cas, as you mentioned below, it could make sense to add that optimization here. This could be worth looking at to see if there's any benefit, but can be done as a possible follow on patch.